src/global_state.cpp

```diff
@@ -159,5 +159,27 @@ void GlobalStateImpl::AddEvent(const std::string& event_type ,
 }
 
 
+void GlobalStateImpl::AddReceiptData(const std::string& data) const {
+    TpReceiptAddDataRequest request;
+    TpReceiptAddDataResponse response;
+
+    request.set_context_id(this->context_id);
+    request.set_data(data);
+
+    FutureMessagePtr future = this->message_stream->SendMessage(
+        Message::TP_RECEIPT_ADD_DATA_REQUEST, request);
+    future->GetMessage(Message::TP_RECEIPT_ADD_DATA_RESPONSE, &response);
+
+    if (response.status() == TpReceiptAddDataResponse::ERROR){
+      std::stringstream error;
+      error << "Failed to add receipt data " << data;
+      throw std::runtime_error(error.str());
```

Comment:
I'm not sure this should be `runtime_error` in either case - in both of these instances, there probably is an issue with the validator.  There's probably an error type that's missing (and this is probably true across the SDK's) that the handler should return, such that it behaves correctly under those circumstances.

Reply:
The python and rust SDKs return INTERNAL_ERROR in the TpProcessResponse status field.
In python that's switched on an InternalError exception. 
In rust there's a specific TransactionReceiptError, but the status field is INTERNAL_ERROR on any error that isn't an invalid transaction, i.e. there's no specific handling for the specific error types.
Probably the best thing for cxx would be to follow the rust SDK pattern and create a new specific exception type in include/exceptions.h 
See the rust SDK for reference: 
https://github.com/hyperledger/sawtooth-sdk-rust/blob/master/src/processor/handler.rs#L68-L80
Specifically `TransactionReceiptError`.
There's already a catchall in the cxx TP process that will operate like the rust process and respond with INTERNAL_ERROR for anything that's not an invalid transaction.

Reply:
@scealiontach where do we sit with this?